### ISSUE TITLE
NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE: False Positive when using Guava

### ISSUE THREAD

**Issue by olenz** [spotbugs/spotbugs#1843]
When using guava in version 31.0.1, some functions seem to cause a false positive in the NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE-checker. 

How to reproduce?
- Use spotbugs (I used 4.5.0, which is the latest version at this moment) on the java class (source code see below).
- To compile it, put guava 31.0.1 on the classpath, which you can get from here: https://mvnrepository.com/artifact/com.google.guava/guava/31.0.1-jre
- Expected behavior: no spotbugs finding.
- Actual behavior: spotbugs will report a "Possible null pointer dereference due to return value of called method" in line 16.

Some remarks:
- When looking at the code of ``linkedHashSetValues()`` (https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/MultimapBuilder.java:354), it is obvious that the method can not actually return a null, therefore I consider this a FP.
- When using guava 27.1 (https://mvnrepository.com/artifact/com.google.guava/guava/27.1-jre), the finding disappears. Note that the code has not significantly changed since then, therefore it must be an issue of the annotations.
- It seems to be associated with the package-info of guava, which uses the annotation com.google.errorprone.annotations.CheckReturnValue (see https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/package-info.java)
- Note, however, that the enclosing class MultimapBuilder has the annotation ElementTypesAreNonnullByDefault, which is intended to cancel the CheckReturnValue annotation (see https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/MultimapBuilder.java and https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/ElementTypesAreNonnullByDefault.java)
- I have tried to reduce the code to not using guava, but I was unable to reproduce the error.

Main.java:
```
package test;

import com.google.common.collect.MultimapBuilder;
import com.google.common.collect.MultimapBuilder.SetMultimapBuilder;

public class Main {
    private enum MyEnum {
        FOOBAR,
    }


    void guava() {
        SetMultimapBuilder builder = MultimapBuilder.enumKeys(MyEnum.class).linkedHashSetValues();

        // The following line will report a NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE
        builder.build();
    }
}
```

**Comment by iloveeclipse** [comment]
I wonder that this all what is needed is the patch below, and if that is semantically correct.

I have no idea what they meant by `"top-level" types`in `Marks all "top-level" types as non-null.`
I guess they mean all non - anonymous types defined in the same class that is annotated?

```
diff --git a/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/NullnessAnnotation.java b/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/NullnessAnnotation.java
index 6c60b72..aa56123 100644
--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/NullnessAnnotation.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/NullnessAnnotation.java
@@ -55,2 +55,3 @@
             return "org.jetbrains.annotations.NotNull".equals(className)
+                    || "com.google.common.collect.ElementTypesAreNonnullByDefault".equals(className)
                     || className.endsWith("Nonnull")
```

**Comment by iloveeclipse** [comment]
> I wonder that all what is needed is the patch below, and if that is semantically correct.

Looking at #182 probably not.

@olenz : do you want to work on the patch ? Tests etc see references PR, should be very similar.

**Comment by olenz** [comment]
I would actually love to work on it, alas, I am afraid I will not find the time for it. :-(

**Comment by josephmate** [comment]
I encountered the same issue but with `MultimapBuilder.hashKeys().arrayListValues().build()` from ` com.google.common.collect.MultimapBuilder`

Could it also be spotbugs gets confused by `@Nullable`in the generic?
```
  public static MultimapBuilderWithKeys<@Nullable Object> hashKeys() {
    return hashKeys(DEFAULT_EXPECTED_KEYS);
  }
```

For now I am working around this issue by adding @SuppressFBWarnings annotations.

**Comment by hgschmie** [comment]
Still present in 4.7.0 ....

**Comment by gtoison** [comment]
This should be fixed by #2502
